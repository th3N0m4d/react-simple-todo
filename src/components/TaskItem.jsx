import React from 'react'
import PropTypes from 'prop-types'
import Button from 'react-bootstrap/Button'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck, faTrash } from '@fortawesome/free-solid-svg-icons'

const propTypes = {
  author: PropTypes.string,
  name: PropTypes.string,
  onConfirm: PropTypes.func,
  onRemove: PropTypes.func
}

const defaultProps = {
  name: 'N/A',
  author: 'Anonymous'
}

const TaskItem = ({
  author,
  name,
  onConfirm,
  onRemove
}) => (
  <li className='list-group-item'>
    <div className='todo-indicator bg-warning' />
    <div className='widget-content p-0'>
      <div className='widget-content-wrapper'>
        <div className='widget-content-left mr-2'>
          <div className='custom-checkbox custom-control'>
            <input className='custom-control-input' id='exampleCustomCheckbox12' type='checkbox' />
            <label className='custom-control-label' htmlFor='exampleCustomCheckbox12'>&nbsp;</label>
          </div>
        </div>
        <div className='widget-content-left'>
          <div className='widget-heading'>
            {name}
          </div>
          <div className='widget-subheading'><i>By {author}</i></div>
        </div>
        <div className='widget-content-right'>
          <Button variant='outline-success' onClick={onConfirm}>
            <FontAwesomeIcon icon={faCheck} />
          </Button>
          <Button variant='outline-danger' onClick={onRemove}>
            <FontAwesomeIcon icon={faTrash} />
          </Button>
        </div>
      </div>
    </div>
  </li>
)

TaskItem.propTypes = propTypes
TaskItem.defaultProps = defaultProps

export default TaskItem
