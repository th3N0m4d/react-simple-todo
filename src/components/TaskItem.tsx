import * as React from 'react'
import Button from 'react-bootstrap/Button'
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome'
import { faCheck, faTrash, faUndo } from '@fortawesome/free-solid-svg-icons'
import cx from 'classnames'

import Variants from '@/constants/Variants'

interface Props {
  author?: string,
  completed?: boolean,
  name?: string,
  onRemove?: () => void,
  onToggle?: () => void,
  variant?: Variants
}

const defaultProps = {
  author: 'Anonymous',
  completed: false,
  name: 'N/A',
  variant: Variants.info,
}

const renderConfirmButton = (onClick: any) => (
  <Button variant='outline-success' onClick={onClick}>
    <FontAwesomeIcon icon={faCheck} />
  </Button>
)

const renderUndoButton = (onClick: any) => (
  <Button variant='outline-info' onClick={onClick}>
    <FontAwesomeIcon icon={faUndo} />
  </Button>
)

const TaskItem: React.FunctionComponent<Props> = ({
  author,
  completed,
  name,
  onRemove,
  onToggle,
  variant
}) => (
    <>
      <div
        className={cx('todo-indicator', {
          [`bg-${variant}`]: true
        })}
      />
      <div className='widget-content p-0'>
        <div className='widget-content-wrapper'>
          <div className='widget-content-left'>
            <div className={cx('widget-heading', {
              'widget-heading--completed': completed
            })}
            >
              {name}
            </div>
            <div className='widget-subheading'><i>By {author}</i></div>
          </div>
          <div className='widget-content-right'>

            {
              completed
                ? renderUndoButton(onToggle)
                : renderConfirmButton(onToggle)
            }
            <Button variant='outline-danger' onClick={onRemove}>
              <FontAwesomeIcon icon={faTrash} />
            </Button>

          </div>
        </div>
      </div>
    </>
  )

TaskItem.defaultProps = defaultProps

export default TaskItem
