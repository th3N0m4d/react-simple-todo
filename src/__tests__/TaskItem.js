import React from 'react'
import { shallow } from 'enzyme'
import renderer from 'react-test-renderer'
import Button from 'react-bootstrap/Button'

import TaskItem from '../components/TaskItem'

const expectToMatchSnapshot = (component) => {
  expect(renderer.create(component).toJSON()).toMatchSnapshot()
}

describe('TaskItem', () => {
  it('should render component correctly', () => {
    expectToMatchSnapshot(<TaskItem />)
  })

  it('should display task name', () => {
    expectToMatchSnapshot(<TaskItem name='Buy wine' />)
  })

  it('should display author\'s name', () => {
    expectToMatchSnapshot(<TaskItem author='Bob' />)
  })

  it('should trigger onRemove callback', () => {
    const onRemoveSpy = jest.fn()
    const INDEX = 1
    const wrapper = shallow(
      <TaskItem onRemove={onRemoveSpy} />
    )

    const button = wrapper.find(Button).at(INDEX)

    button.simulate('click')

    expect(onRemoveSpy).toHaveBeenCalled()
  })

  it('should trigger onConfirm callback', () => {
    const onConfirmSpy = jest.fn()
    const INDEX = 0
    const wrapper = shallow(
      <TaskItem onConfirm={onConfirmSpy} />
    )

    const button = wrapper.find(Button).at(INDEX)

    button.simulate('click')

    expect(onConfirmSpy).toHaveBeenCalled()
  })
})
