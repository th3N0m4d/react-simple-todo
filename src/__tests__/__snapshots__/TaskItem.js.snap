// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TaskItem should display author's name 1`] = `
<li
  className="list-group-item"
>
  <div
    className="todo-indicator bg-warning"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left mr-2"
      >
        <div
          className="custom-checkbox custom-control"
        >
          <input
            className="custom-control-input"
            id="exampleCustomCheckbox12"
            type="checkbox"
          />
          <label
            className="custom-control-label"
            htmlFor="exampleCustomCheckbox12"
          >
             
          </label>
        </div>
      </div>
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
        >
          N/A
        </div>
        <div
          className="widget-subheading"
        >
          <i>
            By 
            Bob
          </i>
        </div>
      </div>
      <div
        className="widget-content-right"
      >
        <button
          className="btn btn-outline-success"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check fa-w-16 "
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <button
          className="btn btn-outline-danger"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash fa-w-14 "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</li>
`;

exports[`TaskItem should display task name 1`] = `
<li
  className="list-group-item"
>
  <div
    className="todo-indicator bg-warning"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left mr-2"
      >
        <div
          className="custom-checkbox custom-control"
        >
          <input
            className="custom-control-input"
            id="exampleCustomCheckbox12"
            type="checkbox"
          />
          <label
            className="custom-control-label"
            htmlFor="exampleCustomCheckbox12"
          >
             
          </label>
        </div>
      </div>
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
        >
          Buy wine
        </div>
        <div
          className="widget-subheading"
        >
          <i>
            By 
            Anonymous
          </i>
        </div>
      </div>
      <div
        className="widget-content-right"
      >
        <button
          className="btn btn-outline-success"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check fa-w-16 "
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <button
          className="btn btn-outline-danger"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash fa-w-14 "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</li>
`;

exports[`TaskItem should render component correctly 1`] = `
<li
  className="list-group-item"
>
  <div
    className="todo-indicator bg-warning"
  />
  <div
    className="widget-content p-0"
  >
    <div
      className="widget-content-wrapper"
    >
      <div
        className="widget-content-left mr-2"
      >
        <div
          className="custom-checkbox custom-control"
        >
          <input
            className="custom-control-input"
            id="exampleCustomCheckbox12"
            type="checkbox"
          />
          <label
            className="custom-control-label"
            htmlFor="exampleCustomCheckbox12"
          >
             
          </label>
        </div>
      </div>
      <div
        className="widget-content-left"
      >
        <div
          className="widget-heading"
        >
          N/A
        </div>
        <div
          className="widget-subheading"
        >
          <i>
            By 
            Anonymous
          </i>
        </div>
      </div>
      <div
        className="widget-content-right"
      >
        <button
          className="btn btn-outline-success"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-check fa-w-16 "
            data-icon="check"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 512 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M173.898 439.404l-166.4-166.4c-9.997-9.997-9.997-26.206 0-36.204l36.203-36.204c9.997-9.998 26.207-9.998 36.204 0L192 312.69 432.095 72.596c9.997-9.997 26.207-9.997 36.204 0l36.203 36.204c9.997 9.997 9.997 26.206 0 36.204l-294.4 294.401c-9.998 9.997-26.207 9.997-36.204-.001z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
        <button
          className="btn btn-outline-danger"
          disabled={false}
          type="button"
        >
          <svg
            aria-hidden="true"
            className="svg-inline--fa fa-trash fa-w-14 "
            data-icon="trash"
            data-prefix="fas"
            focusable="false"
            role="img"
            style={Object {}}
            viewBox="0 0 448 512"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M432 32H312l-9.4-18.7A24 24 0 0 0 281.1 0H166.8a23.72 23.72 0 0 0-21.4 13.3L136 32H16A16 16 0 0 0 0 48v32a16 16 0 0 0 16 16h416a16 16 0 0 0 16-16V48a16 16 0 0 0-16-16zM53.2 467a48 48 0 0 0 47.9 45h245.8a48 48 0 0 0 47.9-45L416 128H32z"
              fill="currentColor"
              style={Object {}}
            />
          </svg>
        </button>
      </div>
    </div>
  </div>
</li>
`;
